### PR TITLE
data workspace extension foundation - batch 2

### PR DESCRIPTION
**Pull request by alanrenmsft**
1. allow other extensions to register themselves as project provider and handles the loading of the project provider extensions if needed
1. update sql-database-project extension to be a project provider extension for 'sqlproj' project type
1. read the projects information from workspace file and populate the projects view
1. add refresh feature for the workspace view
1. interfaces and services - Sakshi needs these for her add project work
1. some unit test cases - more tests are on their way
1. make the menu items available in workspace view
image

in the screenshot below, the project path information are read the workspace file and the sqlproj project type is handled by sql-database-project extension, other extensions can also contribute project types and they will be displayed in this view as well.

image

current test coverage:
image

### CODE REVIEW

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+		ProjectProviderRegistry.providers.forEach(provider => {
+			projectTypes.push(...provider.supportedProjectTypes);
+		});
+		return projectTypes;
+	}
+
+	async getProjectsInWorkspace(): Promise<string[]> {
+		if (vscode.workspace.workspaceFile) {
+			const projects = <string[]>vscode.workspace.getConfiguration(WorkspaceConfigurationName).get(ProjectsConfigurationName);
+			return projects.map(project => path.isAbsolute(project) ? project : path.join(vscode.workspace.rootPath!, project));
```
Comment (Benjin) [inline extensions/data-workspace/src/services/workspaceService.ts:28]: Should (can?) absolute pathing be forced as a convention for registered project providers?  I could see switching back and forth - or forgetting to in places - being a source of bugs.

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	async getProjectProvider(projectFilePath: string): Promise<dataworkspace.IProjectProvider | undefined> {
+		await this.ensureProviderExtensionLoaded(path.extname(projectFilePath));
+		const providers = ProjectProviderRegistry.providers;
+		for (let i = 0; i < providers.length; i++) {
+			if (providers[i].supportedProjectTypes.findIndex(pt => projectFilePath.toUpperCase().endsWith(pt.projectFileExtension.toUpperCase())) !== -1) {
+				return providers[i];
+			}
+		}
```
Comment (Benjin) [inline extensions/data-workspace/src/services/workspaceService.ts:40]: A pair of filter expressions might be cleaner here.

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	async getProjectsInWorkspace(): Promise<string[]> {
+		if (vscode.workspace.workspaceFile) {
+			const projects = <string[]>vscode.workspace.getConfiguration(WorkspaceConfigurationName).get(ProjectsConfigurationName);
+			return projects.map(project => path.isAbsolute(project) ? project : path.join(vscode.workspace.rootPath!, project));
+		}
+		return [];
+	}
+
+	async getProjectProvider(projectFilePath: string): Promise<dataworkspace.IProjectProvider | undefined> {
+		await this.ensureProviderExtensionLoaded(path.extname(projectFilePath));
```
Comment (Benjin) [inline extensions/data-workspace/src/services/workspaceService.ts:34]: What's the behavior if there's no provider registered for a given provider?

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+	async getProjectsInWorkspace(): Promise<string[]> {
+		if (vscode.workspace.workspaceFile) {
+			const projects = <string[]>vscode.workspace.getConfiguration(WorkspaceConfigurationName).get(ProjectsConfigurationName);
+			return projects.map(project => path.isAbsolute(project) ? project : path.join(vscode.workspace.rootPath!, project));
+		}
+		return [];
+	}
+
+	async getProjectProvider(projectFilePath: string): Promise<dataworkspace.IProjectProvider | undefined> {
```
Comment (Benjin) [inline extensions/data-workspace/src/services/workspaceService.ts:33]: If I'm reading the code correctly, this is doing multiple passes through the list of providers that both identify something via a supported project type.  I'm probably being too hopeful here, but is there any way to jump from the vscode extension to the provider so we don't have to scan twice?

File: extensions/sql-database-projects/package.json [hunk 5]
```diff
@@ -234,82 +239,82 @@
       "view/item/context": [
         {
           "command": "sqlDatabaseProjects.build",
-          "when": "view == sqlDatabaseProjectsView && viewItem == databaseProject.itemType.project",
+          "when": "view =~ /^(sqlDatabaseProjectsView|dataworkspace.views.main)$/ && viewItem == databaseProject.itemType.project",
```
Comment (Benjin) [inline extensions/sql-database-projects/package.json:242]: I had no idea we could do this with regex; neat!  Is there any reason to keep sqlDatabaseProjectsView around with the introduction of the dataworkspace view?

File: extensions/sql-database-projects/src/projectProvider/projectPrivider.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (Benjin) [inline extensions/sql-database-projects/src/projectProvider/projectPrivider.ts:1]: Typo in file name: `projectPr*o*vider`

File: extensions/data-workspace/src/common/projectProviderRegistry.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	get providers(): dataworkspace.IProjectProvider[] {
+		return this._providers.slice(0);
+	}
+
+	clear(): void {
+		this._providers.length = 0;
+	}
+
+	validateProvider(provider: dataworkspace.IProjectProvider): void {
+	}
```
[the remark this answers is not in the document]
Comment (alanrenmsft) [reply extensions/data-workspace/src/common/projectProviderRegistry.ts:33]: this is a placeholder, similarly for the workspace service, I will implement the empty methods and add new methods in future PRs

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+const WorkspaceConfigurationName = 'dataworkspace';
+const ProjectsConfigurationName = 'projects';
+
+export class WorkspaceService implements IWorkspaceService {
+	async getAllProjectTypes(): Promise<dataworkspace.IProjectType[]> {
+		await this.ensureProviderExtensionLoaded();
+		const projectTypes: dataworkspace.IProjectType[] = [];
+		ProjectProviderRegistry.providers.forEach(provider => {
+			projectTypes.push(...provider.supportedProjectTypes);
```
[the remark this answers is not in the document]
Comment (alanrenmsft) [reply extensions/data-workspace/src/services/workspaceService.ts:20]: that's what the empty method validateProvider is for :)

[the remark this answers is not in the document]
Comment (alanrenmsft) [reply extensions/data-workspace/src/services/workspaceService.ts:20]: in this extension, I've been using the file extension without the leading dot. later you can clean up the sql-database-project to use the same.

File: extensions/data-workspace/src/services/workspaceService.ts (hunk 1 above)
Comment (alanrenmsft) [reply extensions/data-workspace/src/services/workspaceService.ts:28]: this area needs more attention when I introduce the add project to workspace in the future PR, I will make sure it works.

File: extensions/sql-database-projects/src/projectProvider/projectPrivider.ts (hunk 6 above)
Comment (alanrenmsft) [reply extensions/sql-database-projects/src/projectProvider/projectPrivider.ts:1]: oops!

File: extensions/data-workspace/src/services/workspaceService.ts (hunk 3 above)
Comment (alanrenmsft) [reply extensions/data-workspace/src/services/workspaceService.ts:34]: undefined will be returned and eventually a toast notification will be shown listing the projects with no provider. in workspaceTreeDataProvider

File: extensions/sql-database-projects/package.json (hunk 5 above)
Comment (alanrenmsft) [reply extensions/sql-database-projects/package.json:251]: as we discussed the other day, we will be having these 2 views co-exist for a while.

File: extensions/data-workspace/src/services/workspaceService.ts (hunk 4 above)
Comment (alanrenmsft) [reply extensions/data-workspace/src/services/workspaceService.ts:33]: I've refactored the code a bit and introduced a getProviderByProjectType method on the registry. worst case (the provider extension is not loaded), we still have to call this method twice

**Comment by alanrenmsft** [comment]
not sure why, but this commit history and files are not correct, and the easiest approach to get back to the clean state is to create a new branch and cherry-pick the commits that i made. 

@Benjin I have addressed all the comments in the new branch and sorry for the inconvenience.

**Pull request closed without merging**